### PR TITLE
Added MAiNGO appsi-interface

### PR DESCRIPTION
**Pull request by MAiNGO-github**
## Fixes:
None

## Summary/Motivation:
Added an interface to our global optimization solver MAiNGO

## Changes proposed in this PR:
Add appsi-interface for MAiNGO

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by jsiirola** [comment]
We took a quick look at this during the Dev Call this week.  Most of the test failures are do to a test that tracks the modules that are automatically imported by `pyomo.environ`, and the failure is because the `maingopy` import is getting triggered by
```python
class SolverModel(maingopy.MAiNGOmodel):
```

The easiest workaround is probably to move the `SolverModel` class definition to its own module, and then conditionally import that module into `solvers/maingo.py` (using `attempt_import`).

**Comment by MAiNGO-github** [comment]
> We took a quick look at this during the Dev Call this week. Most of the test failures are do to a test that tracks the modules that are automatically imported by `pyomo.environ`, and the failure is because the `maingopy` import is getting triggered by
> 
> ```python
> class SolverModel(maingopy.MAiNGOmodel):
> ```
> 
> The easiest workaround is probably to move the `SolverModel` class definition to its own module, and then conditionally import that module into `solvers/maingo.py` (using `attempt_import`).

Thank you for the advice, now the import works just fine.

File: pyomo/contrib/appsi/solvers/maingo.py [hunk 1]
```diff
@@ -0,0 +1,482 @@
...
+    ):
+        super(MAiNGOConfig, self).__init__(
+            description=description,
+            doc=doc,
+            implicit=implicit,
+            implicit_domain=implicit_domain,
+            visibility=visibility,
+        )
+
+        self.declare("logfile", ConfigValue(domain=str))
```
Comment (mrmundt) [inline pyomo/contrib/appsi/solvers/maingo.py:90]: You can specify a default here and remove L#94:
```python
self.declare("logfile", ConfigValue(domain=str, default=None)
```

File: pyomo/contrib/appsi/solvers/maingo.py [hunk 2]
```diff
@@ -0,0 +1,482 @@
...
+        self.declare("solver_output_logger", ConfigValue())
```
Comment (mrmundt) [inline pyomo/contrib/appsi/solvers/maingo.py:91]: You can specify a default here and remove L#95:
```python
self.declare("solver_output_logger", ConfigValue(default=logger)
```

File: pyomo/contrib/appsi/solvers/maingo.py [hunk 3]
```diff
@@ -0,0 +1,482 @@
...
+        self.declare("log_level", ConfigValue(domain=NonNegativeInt))
```
Comment (mrmundt) [inline pyomo/contrib/appsi/solvers/maingo.py:92]: You can specify a default here and remove L#96:
```python
self.declare("log_level", ConfigValue(domain=NonNegativeInt, default=20)
```

File: pyomo/contrib/appsi/solvers/maingo.py [hunk 4]
```diff
@@ -0,0 +1,482 @@
...
+        pass
+
+    def _remove_params(self, params: List[_ParamData]):
+        pass
+
+    def _update_variables(self, variables: List[_GeneralVarData]):
+        pass
+
+    def update_params(self):
+        pass
```
Comment (mrmundt) [inline pyomo/contrib/appsi/solvers/maingo.py:360]: Why pass all of these?

File: pyomo/contrib/appsi/solvers/maingo.py [hunk 5]
```diff
@@ -0,0 +1,482 @@
...
+        status = mprob.get_status()
+        results = MAiNGOResults(solver=self)
+        results.wallclock_time = mprob.get_wallclock_solution_time()
+        results.cpu_time = mprob.get_cpu_solution_time()
+
+        if status in {maingopy.GLOBALLY_OPTIMAL, maingopy.FEASIBLE_POINT}:
+            results.termination_condition = TerminationCondition.optimal
+            if status == maingopy.FEASIBLE_POINT:
+                logger.warning(
+                    "MAiNGO did only find a feasible solution but did not prove its global optimality."
```
Comment (mrmundt) [inline pyomo/contrib/appsi/solvers/maingo.py:387]:
```suggestion
                    "MAiNGO found a feasible solution but did not prove its global optimality."
```

**Review by mrmundt — CHANGES_REQUESTED** [review]
Thanks for this! I did an initial pass and have some comments. Also, can you please add our standard copyright statement to all new files? It can be found in any other existing `*.py` file.

File: pyomo/contrib/appsi/solvers/maingo.py (hunk 1 above)
Comment (MAiNGO-github) [reply pyomo/contrib/appsi/solvers/maingo.py]: Thank you for the tip, done.

File: pyomo/contrib/appsi/solvers/maingo.py (hunk 3 above)
Comment (MAiNGO-github) [reply pyomo/contrib/appsi/solvers/maingo.py]: done.

File: pyomo/contrib/appsi/solvers/maingo.py [hunk 6]
```diff
@@ -0,0 +1,482 @@
...
+        if sys.version_info.major >= 3 and sys.version_info.minor >= 8:
+            from importlib.metadata import version
+
+            version = version('maingopy')
+        else:
+            import pkg_resources
+
+            version = pkg_resources.get_distribution('maingopy').version
+
+        return tuple(int(k) for k in version.split('.'))
```
[the remark this answers is not in the document]
Comment (MAiNGO-github) [reply pyomo/contrib/appsi/solvers/maingo.py]: Removed.

File: pyomo/contrib/appsi/solvers/maingo.py (hunk 4 above)
Comment (MAiNGO-github) [reply pyomo/contrib/appsi/solvers/maingo.py]: We added the missing functionality and no longer pass these functions.

**Comment by MAiNGO-github** [comment]
Dear @mrmundt 
We have adressed all your comments and fixed some bugs. The interface does no pass all tests (locally on our machine).

There are a few points we need your input on:
- MAiNGO does not compute duals or reduced costs. We have extended the tests, so that duals or RCs are not computed for MAiNGO. Maybe there is a nicer way to do this?
- MAiNGO has rather loose tolerances by default. To pass the tests we needed to raise them to a level which we do not want to set as default. For now, we create a new solverclass that inherits from our base solver class that uses tighter tolerances (MAiNGOTest). Again, is there a nicer way to do this?
- MAiNGO does not support unbounded variables. To deal with unbounded variables we bound them to [-1e8,1e8]. As a consequence, we cannot recognice a problem to be unbounded.
- We had to modify two tests (make variables > 0) as MAiNGO cannot handle log of negative values or division by 0.

Thank you!

**Comment by mrmundt** [comment]
@MAiNGO-github - I will reply to your other comments soon, but FYI on the two failures:
1. `conda` is not your fault; something just didn't download correctly
2. `singletest` is looking to make sure that you have registered the SolverFactory in the plugins system. See here: https://github.com/Pyomo/pyomo/blob/main/pyomo/contrib/appsi/plugins.py <- you'll need to add your new solver to the list.

File: pyomo/contrib/appsi/solvers/tests/test_persistent_solvers.py [hunk 7]
```diff
@@ -1359,7 +1390,8 @@ def test_param_updates(self, name: str, opt_class: Type[PersistentSolver]):
         m.obj = pe.Objective(expr=m.y)
         m.c1 = pe.Constraint(expr=(0, m.y - m.a1 * m.x - m.b1, None))
         m.c2 = pe.Constraint(expr=(None, -m.y + m.a2 * m.x + m.b2, 0))
-        m.dual = pe.Suffix(direction=pe.Suffix.IMPORT)
+        if not opt_class is MAiNGO:
```
Comment (mrmundt) [inline pyomo/contrib/appsi/solvers/tests/test_persistent_solvers.py:1393]: This is fine for now. We have an open backlog item to redesign the solver tests from scratch, so we can try to find a more clever way to do this in the future.

One request, though - universally change the check to:
```
if opt_class != MAiNGO
```
(Assuming that doesn't error for some Pyomo-related reason...)

**Review by mrmundt — CHANGES_REQUESTED** [review]
Addressed your questions!

**Comment by mrmundt** [comment]
@MAiNGO-github - This is insidious, and you would never have known it happened. `maingopy` isn't installing correctly:
```
Collecting maingopy
  Downloading maingopy-0.7.3.tar.gz (17.9 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 17.9/17.9 MB 75.6 MB/s eta 0:00:00
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'error'
  error: subprocess-exited-with-error
  
  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> [13 lines of output]
      Traceback (most recent call last):
        File "/tmp/pip-install-du9lfhn7/maingopy_85dd9705743e442785b6d05edb86f03d/setup.py", line 7, in <module>
          from skbuild import setup
      ModuleNotFoundError: No module named 'skbuild'
      
      During handling of the above exception, another exception occurred:
      
      Traceback (most recent call last):
        File "<string>", line 2, in <module>
        File "<pip-setuptools-caller>", line 34, in <module>
        File "/tmp/pip-install-du9lfhn7/maingopy_85dd9705743e442785b6d05edb86f03d/setup.py", line 9, in <module>
Error:           raise ImportError('Error during setup of maingopy: maingopy requires scikit-build, which is not available.')
      ImportError: Error during setup of maingopy: maingopy requires scikit-build, which is not available.
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
WARNING: MAiNGO is not available
```

File: pyomo/contrib/appsi/solvers/tests/test_persistent_solvers.py [hunk 8]
```diff
@@ -1217,16 +1247,16 @@ def test_with_gdp(
...
 
         opt: PersistentSolver = opt_class(only_child_vars=only_child_vars)
         opt.use_extensions = True
         res = opt.solve(m)
-        self.assertAlmostEqual(res.best_feasible_objective, 1)
-        self.assertAlmostEqual(m.x.value, 0)
-        self.assertAlmostEqual(m.y.value, 1)
+        self.assertAlmostEqual(res.best_feasible_objective, 1, 6)
+        self.assertAlmostEqual(m.x.value, 0, 6)
+        self.assertAlmostEqual(m.y.value, 1, 6)
```
Comment (mrmundt) [inline pyomo/contrib/appsi/solvers/tests/test_persistent_solvers.py:1259]: Does the default (`places=7`) cause failures?

Comment (MAiNGO-github) [reply pyomo/contrib/appsi/solvers/tests/test_persistent_solvers.py:1259]: With the default tolerances set in maingo.py,  `places=7` causes failures. We could tighten the tolerances even further for the tests. But we don't want to raise the default tolerances.

**Review by mrmundt — APPROVED** [review]

**Review by blnicho — APPROVED** [review]
I think this looks great. I have one minor change request, could you add maingo to the list of appsi solvers in this documentation page: https://github.com/Pyomo/pyomo/blob/main/doc/OnlineDocs/library_reference/appsi/appsi.solvers.rst

**Comment by mrmundt** [comment]
> I think this looks great. I have one minor change request, could you add maingo to the list of appsi solvers in this documentation page: https://github.com/Pyomo/pyomo/blob/main/doc/OnlineDocs/library_reference/appsi/appsi.solvers.rst

Added in last two commits.

**Pull request merged**